File: script/import.py
# Use python2
# Data
# Use this class to collect all the scattered csv file data into
# a format with can use later tht matches our database
#
# Authors:
# Jacob Gonzalez
# .

import csv

# course classthat fits the ruby model
class Course:
    def __init__(self, id, name, catalog_number):
        self.id = id
        self.name = name
        self.catalog_number = catalog_number
        self.components = []

    # print as a csv row
    def __repr__(self):
        return '"%s","%s","%s"' % (self.id, self.name, self.catalog_number)

    # test equality against another course
    def __eq__(self, other):
        return self.id == other.id

# component links courses and sessions
# e.g. LEC, TUT, etc
class Component:
    def __init__(self, id, type):
        self.id = id
        self.type = type

    def __repr__(self):
        return '"%s"' % self.type

    def __eq__(self, other):
        return self.type == other.type

# import courses from the course catalog
def importCourses(filename):
    courses = {}
    with open(filename) as f:
        reader = csv.reader(f)
        # each row is a new course
        for row in reader:
            id = row[1]
            name = row[4]
            course = Course(id, name, None)
            courses[id] = course
    return courses

# fill offering cataglog numbers from the offerings csv
def fillCourseOfferings(filename, courses):
    with open(filename) as f:
        reader = csv.reader(f)
        for row in reader:
            id = row[1]
            if id in courses:
                # COMPSCI 3005
                courses[id].catalog_number = '%s %s' % (row[4], row[5])

# read in all the components and link them to their courses
def importComponentsAndLink(filename, courses):
    components = []
    with open(filename) as f:
        reader = csv.reader(f)
        for row in reader:
            id = row[1]
            type = row[-1]
            component = Component(id, type)
            # link component and course
            if id in courses:
                course = courses[id]
                if component not in course.components:
                    courses[id].components.append(component)
                    components.append(component)

    return components

directory = raw_input('directory? ')

# TODO: find a bettr place for the course csv files
courses = importCourses('%s/CM_CRSE_CAT_ECMS-6383074.csv' % directory)
fillCourseOfferings('%s/CM_CRSE_CAT_ECMS_OFFERINGS-6383075.csv' % directory, courses)
components = importComponentsAndLink('%s/CM_CRSE_CAT_ECMS_COMPONENTS-6383069.csv' % directory, courses)

# print them for now
for course in courses.itervalues():
    print(course)
    print('\t%s' % course.components)

File: script/import.rb
# ported to ruby by: Jacob Gonzalez
# Use this class to collect all the scattered csv file data into
# a format with can use later tht matches our database
#
# Authors:
# Jacob Gonzalez
# .

require 'csv'

# course class that fits the ruby model
# replace this with the ruby model
class Course
    attr_accessor :id
    attr_accessor :name
    attr_accessor :catalog_number
    attr_accessor :components

    def initialize(id, name, catalog_number)
        @id = id
        @name = name
        @catalog_number = catalog_number
        @components = []
    end
end

# component links courses and sessions
# e.g. LEC, TUT, etc
class Component
    attr_accessor :id
    attr_accessor :type

    def initialize(id, type)
        @id = id
        @type = type
    end
end

# import courses from the course catalog
def importCourses(filename)
    courses = {}
    # each row is a new course
    CSV.foreach(filename, headers: true, encoding: 'iso-8859-1:utf-8') do |row|
        id = row[1]
        name = row[4]
        course = Course.new(id, name, "")
        courses[id] = course
    end
    return courses
end

# fill offering cataglog numbers from the offerings csv
def fillCourseOfferings(filename, courses)
    CSV.foreach(filename, headers: true, encoding: 'iso-8859-1:utf-8') do |row|
        id = row[1]
        if courses.has_key? id
            # e.g. COMPSCI 3005
            courses[id].catalog_number = "#{row[4]} #{row[5]}"
        end
    end
end

# read in all the components and link them to their courses
def importComponentsAndLink(filename, courses)
    components = []
    CSV.foreach(filename, headers: true, encoding: 'iso-8859-1:utf-8') do |row|
        id = row[1]
        type = row[-1]
        component = Component.new(id, type)
        if courses.has_key? id
            course = courses[id]
            # e.g. COMPSCI 3005
            unless course.components.include? component
                course.components.push component
                components.push component
            end
        end
    end
    return components
end

# read and import the files
directory = 'db/csv'
courses = importCourses "#{directory}/CM_CRSE_CAT_ECMS-6383074.csv"
fillCourseOfferings "#{directory}/CM_CRSE_CAT_ECMS_OFFERINGS-6383075.csv", courses
components = importComponentsAndLink("#{directory}/CM_CRSE_CAT_ECMS_COMPONENTS-6383069.csv", courses)

# print them for now
courses.each do |id, course|
    puts "#{id},#{course.name},#{course.catalog_number}"
    puts "\t#{course.components.map{|c| c.type}.join(',')}"
end
